tools/im2rec.py

```diff
@@ -1,238 +1,265 @@
+        from the input list if >1. the input list will be modified to match the\
+        resulting order.')
+    rgroup.add_argument('--color', type=int, default=1, choices=[-1, 0, 1],
+                        help='specify the color mode of the loaded image.\
+        1: Loads a color image. Any transparency of image will be neglected. It is the default flag.\
+        0: Loads image in grayscale mode.\
+        -1:Loads image as such including alpha channel.')
+    rgroup.add_argument('--encoding', type=str, default='.jpg', choices=['.jpg', '.png'],
+                        help='specify the encoding of the images.')
+    rgroup.add_argument('--saving-folder', type=str, default='.',
+                        help='folder in which .rec files will be saved.')
+    rgroup.add_argument('--shuffle', default=True, help='If this is set as True, \
+        im2rec will randomize the image order in <prefix>.lst')
+    args = parser.parse_args()
+
+    # add data_path here
+    # data_path = os.path.abspath(args.root)
+    data_path = args.root
+    if args.list:
+        make_list(args)
+    else:
+        # f is just file name original, not a path, here changed it
+        files = [f for f in os.listdir(data_path) if os.path.isfile(os.path.join(data_path, f))]
+        for f in files:
+            if f.startswith(args.prefix) is True and f.endswith('.lst') is True:
+                print 'Creating .rec file from', f, 'in', args.saving_folder
+                # join f with data_path
+                image_list = read_list(os.path.join(data_path, f))
+                # delete write record and moved it to the __main__ process
+                return (args, image_list, f)
+
+
+if __name__ == '__main__':
+    # here is the __main__ process, which do the write operation original in write_record 
+    (args, image_list, fname) = main()
+    source = image_list
+    tic = [time.time()]
+    try:
+        import multiprocessing
+        # add freeze_support
+        multiprocessing.freeze_support()
+        q_in = [multiprocessing.Queue() for i in range(args.num_thread)]
+        q_out = multiprocessing.Queue(1024)
+        for i in range(len(image_list)):
+            q_in[i % len(q_in)].put(image_list[i])
+        read_process = [multiprocessing.Process(target=read_worker, args=(args, q_in[i], q_out)) \
+                        for i in range(args.num_thread)]
+        for p in read_process:
+            p.start()
+        write_process = multiprocessing.Process(target=write_worker, args=(q_out, fname, args.saving_folder))
+        write_process.start()
+        for p in read_process:
+            p.join()
+        q_out.put(('finish', '', []))
+        write_process.join()
+    except EOFError:
+        print('multiprocessing not available, fall back to single threaded encoding')
+        import Queue
+        q_out = Queue.Queue()
+        fname_rec = fname[:fname.rfind('.')]

```

Comment:
remove the change directory operation, change the write record operation controled by write function it self